src/mbgl/actor/thread_pool.cpp

```diff
@@ -0,0 +1,48 @@
+#include <mbgl/actor/thread_pool.hpp>
+#include <mbgl/actor/mailbox.hpp>
+
+namespace mbgl {
+
+ThreadPool::ThreadPool(std::size_t count) {
+    threads.reserve(count);
+    for (std::size_t i = 0; i < count; ++i) {
+        threads.emplace_back([this] () {
+            while (true) {
+                std::unique_lock<std::mutex> lock(mutex);
+
+                cv.wait(lock, [this] {
+                    return !queue.empty() || terminate.load();
+                });
+
+                if (terminate.load()) {
+                    return;
+                }
+
+                auto mailbox = queue.front();
+                queue.pop();
+                lock.unlock();
+
+                if (auto locked = mailbox.lock()) {
```

Comment:
This is the idiomatic use of `weak_ptr`; see e.g. [here](http://en.cppreference.com/w/cpp/memory/weak_ptr). In general, declaring on a separate line would require an extra set of braces and indentation to preserve the same region of locking:

```
{
    auto locked = mailbox.lock();
    if (locked) {
        ...
    }
}
```